File: codebase/core/main/src/Engine/UI/PhysicalViewDomProcessor.fs
﻿namespace Forest.UI

open System
open Forest
open Forest.Engine
open Axle


[<NoComparison;StructuralEquality>] 
type internal NodeState =
    | NewNode of node : DomNode
    | UpdatedNode of node : DomNode

[<Sealed;NoComparison;NoEquality>] 
type internal PhysicalViewDomProcessor =
    val mutable private physicalViews : Map<thash, IPhysicalView>
    /// Contains a list of nodes to be retained as they represent present views.
    val mutable private _nodesToPreserve : thash Set
    val mutable private _nodeStates : NodeState list
    val private _engine : IForestEngine
    val private _renderer : IPhysicalViewRenderer

    private new (physicalViews, engine : IForestEngine, renderer : IPhysicalViewRenderer) = 
        { 
            _engine = engine; 
            _renderer = renderer; 
            physicalViews = physicalViews; 
            _nodesToPreserve = Set.empty; 
            _nodeStates = List.empty 
        }
    new (engine : IForestEngine, renderer : IPhysicalViewRenderer) = PhysicalViewDomProcessor(Map.empty, engine, renderer)

    member this.PhysicalViews 
        with get() = this.physicalViews
        and set pv = this.physicalViews <- pv

    interface IDomProcessor with
        member this.ProcessNode n =
            this._nodesToPreserve <- this._nodesToPreserve |> Set.add n.InstanceID
            this._nodeStates <- 
                match this.physicalViews.TryFind n.InstanceID with
                | Some _ -> (UpdatedNode n)::this._nodeStates
                | None -> (NewNode n)::this._nodeStates
            n

        member this.Complete(_ : DomNode seq) = 
            let physicalViews = System.Collections.Generic.Dictionary<thash, IPhysicalView>(StringComparer.Ordinal)
            for kvp in this.physicalViews do
                if not <| this._nodesToPreserve.Contains(kvp.Key) then
                    kvp.Value.Dispose()
                else
                    physicalViews.Add(kvp.Key, kvp.Value)

            let inline dictTryFind (dict : System.Collections.Generic.Dictionary<thash, 'PV>, key) =
                match dict.TryGetValue key with
                | (false, _) -> None
                | (true, value) -> Some value

            // store tuples of renderer and node to initiate an update call after the views are rendered
            let mutable updateCallArguments = List.empty

            for nodeState in this._nodeStates do
                let (n, p, isNewView) = 
                    match nodeState with 
                    | NewNode n -> (n, n.Parent, true) 
                    | UpdatedNode n -> (n, n.Parent, false)
                match (null2opt p, isNewView) with
                | (Some p, isNewView) -> 
                    match (isNewView, dictTryFind(physicalViews, p.InstanceID), dictTryFind(physicalViews, n.InstanceID)) with
                    | (true, Some parent, None) ->
                        let physicalView = this._renderer.CreateNestedPhysicalView (this._engine, parent, n)
                        updateCallArguments <- (physicalView,n)::updateCallArguments
                        physicalViews.Add(n.InstanceID, physicalView)
                    | (true, Some _, Some _) ->
                        invalidOp(String.Format("Expecting physical view {0} #{1} not to contain child {2} #{3}", p.Name, p.InstanceID, n.Name, n.InstanceID))
                    | (false, Some _, Some renderer) ->
                        updateCallArguments <- (renderer,n)::updateCallArguments
                    | (false, Some _, None) ->
                        invalidOp(String.Format("Could not locate physical view {0} #{1}", n.Name, n.InstanceID))
                    | (_, None, _) ->
                        invalidOp(String.Format("Could not locate physical view {0} #{1} that should be parent of {2} #{3}", p.Name, p.InstanceID, n.Name, n.InstanceID))
                | (None, isNewView) ->
                    match (isNewView, dictTryFind(physicalViews, n.InstanceID)) with
                    | (true, Some _) ->
                        invalidOp(String.Format("Physical view {0} #{1} already exists", n.Name, n.InstanceID))
                    | (true, None) ->
                        let physicalView = this._renderer.CreatePhysicalView(this._engine, n)
                        updateCallArguments <- (physicalView,n)::updateCallArguments
                        physicalViews.Add(n.InstanceID, physicalView)
                    | (false, None) -> 
                        invalidOp(String.Format("Could not locate physical view {0} #{1}", n.Name, n.InstanceID))
                    | (false, Some renderer) ->
                        updateCallArguments <- (renderer,n)::updateCallArguments

            this.physicalViews <- physicalViews |> Seq.map (|KeyValue|) |> Map.ofSeq
            this._nodesToPreserve <- Set.empty
            this._nodeStates <- List.empty

            // Initiate the update calls for renderers. This must happen after the renderers collections is up to date
            // because an update call might trigger an Forest command and case re-rendering before the rendered views are processed.
            // The consequences of a command execution during the physical UI construction could result in attempt to render the 
            // same physical view twice, thus causing an exception.
            // At this point, invoking those update calls is permitted.
            for (renderer,n) in updateCallArguments do 
                renderer.Update n

File: codebase/engine/main/src/UI/PhysicalViewDomProcessor.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using Forest.Engine;

namespace Forest.UI
{
    internal sealed class PhysicalViewDomProcessor : IDomProcessor
    {
        public enum NodeState : sbyte
        {
            NewNode = 0,
            UpdatedNode = 1
        }

        private IImmutableSet<string> _nodesToPreserve = ImmutableHashSet<string>.Empty;
        private IImmutableList<Tuple<DomNode, NodeState>> _nodeStates = ImmutableList<Tuple<DomNode, NodeState>>.Empty;

        private readonly IForestEngine _engine;
        private readonly IPhysicalViewRenderer _renderer;

        private PhysicalViewDomProcessor(IForestEngine engine, IPhysicalViewRenderer renderer, StringComparer stringComparer, ImmutableDictionary<string, IPhysicalView> physicalViews)
        {
            _engine = engine;
            _renderer = renderer;
            PhysicalViews = physicalViews == null
                ? ImmutableDictionary.Create<string, IPhysicalView>(StringComparer.Ordinal)
                : ImmutableDictionary.CreateRange(physicalViews.KeyComparer, physicalViews);
        }
        internal PhysicalViewDomProcessor(IForestEngine engine, IPhysicalViewRenderer renderer, ImmutableDictionary<string, IPhysicalView> physicalViews = null)
            : this(engine, renderer, StringComparer.Ordinal, physicalViews) { }
        public PhysicalViewDomProcessor(IForestEngine engine, IPhysicalViewRenderer renderer) 
            : this(engine, renderer, null) { }

        [Obsolete]
        public ImmutableDictionary<string, IPhysicalView> PhysicalViews { get; internal set; }

        DomNode IDomProcessor.ProcessNode(DomNode node)
        {
            _nodesToPreserve = _nodesToPreserve.Add(node.InstanceID);
            _nodeStates = _nodeStates.Add(
                Tuple.Create(
                    node, 
                    PhysicalViews.TryGetValue(node.InstanceID, out _) ? NodeState.UpdatedNode : NodeState.NewNode));
            return node;
        }

        void IDomProcessor.Complete(IEnumerable<DomNode> nodes)
        {
            var physicalViews = new Dictionary<string, IPhysicalView>(PhysicalViews.KeyComparer);
            foreach (var kvp in PhysicalViews)
            {
                if (_nodesToPreserve.Contains(kvp.Key))
                {
                    physicalViews.Add(kvp.Key, kvp.Value);
                }
                else
                {
                    //
                    // The respective logical view was removed and is no longer present in the state.
                    //
                    kvp.Value.Dispose();
                }
            }

            // store tuples of renderer and node to initiate an update call after the views are rendered
            var updateCallArguments = new List<Tuple<IPhysicalView, DomNode>>();
            foreach(var nodeState in _nodeStates)
            {
                var n = nodeState.Item1;
                var isNewView = nodeState.Item2 == NodeState.NewNode;
                var current = physicalViews.TryGetValue(n.InstanceID, out var _n) ? _n : null;

                IPhysicalView parent = null;
                if (n.Parent == null || (n.Parent != null && physicalViews.TryGetValue(n.Parent.InstanceID, out parent)))
                {
                    if (current == null && isNewView)
                    {
                        var physicalView = parent != null
                            ? _renderer.CreateNestedPhysicalView(_engine, parent, n)
                            : _renderer.CreatePhysicalView(_engine, n);
                        //updateCallArguments < -(physicalView, n)::updateCallArguments
                        updateCallArguments.Add(Tuple.Create(physicalView, n));
                        physicalViews.Add(n.InstanceID, physicalView);
                    }
                    else if (current != null && !isNewView)
                    {
                        updateCallArguments.Add(Tuple.Create(current, n));
                    }
                    else if (current != null && isNewView)
                    {
                        throw new InvalidOperationException(n.Parent != null
                            ? string.Format("Did not expect physical view {0} #{1} to contain child {2} #{3}", n.Parent.Name, n.Parent.InstanceID, n.Name, n.InstanceID)
                            : string.Format("Physical view {0} #{1} already exists", n.Name, n.InstanceID));
                    }
                    else if (current == null && !isNewView)
                    {
                        throw new InvalidOperationException(string.Format("Could not locate physical view {0} #{1}", n.Name, n.InstanceID));
                    }
                }
                else if (n.Parent != null && parent == null)
                {
                    throw new InvalidOperationException(string.Format("Could not locate physical view {0} #{1} that should be parent of {2} #{3}", n.Parent.Name, n.Parent.InstanceID, n.Name, n.InstanceID));
                }
            }

            PhysicalViews = ImmutableDictionary.CreateRange(physicalViews.Comparer, physicalViews);
            _nodesToPreserve = _nodesToPreserve.Clear();
            _nodeStates = _nodeStates.Clear();

            // Initiate the update calls for the enumerated physical views.
            //
            // Note that this happens AFTER all physical views have been collected.
            // If we attempt to call Update on a physical view without being done with the above rendering,
            // we risk that the call might trigger a Forest command or message passing instruction,
            // which in turn would cause a nested render call within the current render, while it is still operating.
            //
            // The consequences of the above situation could result in rendering existing physical view twice,
            // which has unknown consequences.
            // In addition, only the results of the outermost render call will be respected,
            // meaning any views created in the nested call(s) will not be displayed, while still residing in the memory.
            // 
            // At this point, invoking those update calls is safe and the described above side-effects are avoided.
            foreach (var x in updateCallArguments)
            {
                x.Item1.Update(x.Item2);
            }
        }
    }
}
